### DISCUSSION TITLE
How to create the App Package with the WinUi Desktop mode using cake plugin

### DISCUSSION THREAD

**Discussion by DhanushKRaja** [cake-build/cake#3411] [Q&A]
In the build system, I'm trying to create a App Package with the WinUi Desktop mode using cake plugins. I could create this same for UWP using the below msbuild property. But, I could not create the App Package for WinUi Desktop mode.

I am using below configuration in my build systems.

MSBuildToolVersion VS2019 (tried with Default).
SetConfiguration: Release-Xml (tried with Release).

Please find the below MsBuild() which is working fine foe UWP App Package creation. I have used the same MsBuild() for WinUi Desktop mode. But, I could not achieve by this way.

context.MSBuild(csProject.FullName, settings =>
                        settings.SetConfiguration(configuration).
                        UseToolVersion(msBuildVersion).
                        WithProperty("referencepath", referencepath).
                        WithProperty("PreBuildEvent", "").
                        WithProperty("GenerateLibraryLayout", "true").
                        AddFileLogger(new MSBuildFileLogger { LogFile = warningLog, MSBuildFileLoggerOutput = MSBuildFileLoggerOutput.WarningsOnly }).
                        AddFileLogger(new MSBuildFileLogger { LogFile = errorLog, MSBuildFileLoggerOutput = MSBuildFileLoggerOutput.ErrorsOnly }).
                        WithProperty("CodeAnalysisRuleSet", $"{csProject.Directory}/FxCop.ruleset").
                        WithProperty("StyleCopOverrideSettingsFile", $"{csProject.Directory}/settings.StyleCop").
                        WithProperty("TreatWarningsAsErrors", assembly.TreatWarningsAsErrors.ToString()).
                        WithProperty("OutputPath", projoutputpath).
                        WithProperty("IntermediateOutputPath", intermediateOutputPath).
                        SetVerbosity(Cake.Core.Diagnostics.Verbosity.Minimal));

Visual Studio Configuration: 

- Visual Studio 2019 v16.10.3 

Please suggest how to create a App Package for WinUi Desktop mode using Cake Plugins.

**Comment by patriksvensson** [comment]
I would suggest that you try getting it work by invoking MSBuild from the command line first. After that, I'm sure there are people who can help you convert that into a Cake script.

**Reply by DhanushKRaja** [reply]
Hi @patriksvensson 

Thanks for your suggestion.

I have tried to execute the app Package creation by the below command line.

msbuild "AppTest.sln" /p:Configuration=Release;Platform="x86";OutputPath=C:\AppPackage;UapAppxPackageBuildMode=StoreUpload /m
Using Platform="x86" instead of Release="Windows-x86".

But, I am facing the below Error messages. Can I have any other suggestion for resolve it.

![Buildfailure](https://user-images.githubusercontent.com/87820212/127097231-0107e192-b003-4b2e-959a-fb5aa93707f2.jpg)

Regards,
DhanushKRaja

**Reply by nils-a** [reply]
@DhanushKRaja the error seems to be

> The specified solution configuration "Release|ARM" is invalid...

So, I'm guessing either your argument `/property:Configuration="Release"` or your argument `/p:Platform="ARM"` does not match the configuration of the solution.

**Reply by DhanushKRaja** [reply]
Hi @nils-a 

Thanks for take look on this issue.

### **This is my requirement** :

In WinUI platform there is 2 type of solution we could use one is **UWP solution** another one is **Desktop solution**.
I can create a App Package for  **UWP solution**(which hold's simply solution.winui) By the below msbuild code.

```
context.MSBuild(csProject.FullName, settings =>
                                settings.SetConfiguration("Release").
                                UseToolVersion("VS2019").
                                WithProperty("SetMSBuildPlatform", "AnyCPU").
                                WithProperty("referencepath", referencepath).
                                WithRestore().
                                WithTarget("Build").
                                WithProperty("PreBuildEvent", "").
                                WithProperty("GenerateLibraryLayout", "true").
                                AddFileLogger(new MSBuildFileLogger { LogFile = warningLog, MSBuildFileLoggerOutput = MSBuildFileLoggerOutput.WarningsOnly }).
                                AddFileLogger(new MSBuildFileLogger { LogFile = errorLog, MSBuildFileLoggerOutput = MSBuildFileLoggerOutput.ErrorsOnly }).
                                WithProperty("WarningLevel", assembly.WarningLevel.ToString()).
                                WithProperty("TreatWarningsAsErrors", assembly.TreatWarningsAsErrors.ToString()).
                                WithProperty("OutputPath", projoutputpath).
                                WithProperty("DefineConstants", defineConstants).
                                WithProperty("IntermediateOutputPath", intermediateOutputPath).
                                WithProperty("AppxBundlePlatforms", "x86").
                                WithProperty("AppxBundle", "Always").
                                WithProperty("UapAppxPackageBuildMode", "Always").
                                WithProperty("UapAppxPackageBuildMode", "SideloadOnly").
                                WithProperty("AppxSymbolPackageEnabled", "True").
                                WithProperty("AppxPackageSigningEnabled", "True").
                                WithProperty("PackageCertificateThumbprint", "2DA1648AA2E87F9B2805").
                                WithProperty("PackageCertificateKeyFile", $"{csProject.Directory}/project.pfx").
                                WithProperty("PackageCertificatePassword", "password").
                                SetVerbosity(Cake.Core.Diagnostics.Verbosity.Minimal));
```

In the same I am trying to get the App Package for Desktop solution.  When I execute the same code for create an App package for Desktop solution, I could not achieve my target.

When I use the above code for create the App Package for UWP solution, It working fine. Then, I am using the same set of code for execute the Desktop Solution(Contain(winui_net50)), Its not working As Iike  **UWP solution**.

Can you please suggest me any solution to resolve this.

Regards,
DhanushKRaja

**Reply by nils-a** [reply]
@DhanushKRaja I'm not sure I can follow - but I don't build UWP or WinUI. 

As @patriksvensson suggested I would keep things simple and try using simple `msbuild` commandline statements. If you know what `msbuild` commands build your solution, we can help you get those into a Cake build.
